Currently the typings for mini-css-extract-plugin and duplicate-package-checker-webpack-plugin are written for the DefinitelyTyped Webpack 4 typings. Wepback 5 comes with its own typings which are not compatible. Since this branch switches us to Webpack 5, we manually patch the typings for these plugins to be compatible with the Webpack 5 typings.

To use, do a `pip install -ve .` (which essentially does a yarn install) to get the node_modules directory and download the current typings for these webpack plugins. This will give some compilation/typing errors like the following:

```
src/build.ts(196,9): error TS2322: Type 'MiniCssExtractPlugin[]' is not assignable to type '(WebpackPluginInstance | ((this: Compiler, compiler: Compiler) => void))[]'.
```

To fix these, do the following in the repo root directory:

```
patch -p1 < webpack5-types.patch
```

to apply these changes to the node_modules directory. Then you can do `pip install -ve .` again to finish the installation process.

--- a/node_modules/@types/mini-css-extract-plugin/index.d.ts	2020-05-26 12:33:41.000000000 -0700
+++ b/node_modules/@types/mini-css-extract-plugin/index.d.ts	2020-05-26 14:52:08.000000000 -0700
@@ -6,18 +6,20 @@
 //                 Piotr Błażejewicz <https://github.com/peterblazejewicz>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-import { compilation, Plugin } from 'webpack';
+import { Chunk, Compiler } from 'webpack';
 
 /**
  * Lightweight CSS extraction webpack plugin
  * This plugin extract CSS into separate files. It creates a CSS file per JS file which contains CSS. It supports On-Demand-Loading of CSS and SourceMaps.
  * Configuration Detail: https://github.com/webpack-contrib/mini-css-extract-plugin#configuration
  */
-declare class MiniCssExtractPlugin extends Plugin {
+declare class MiniCssExtractPlugin {
     /** webpack loader used always at the end of loaders list */
     static loader: string;
 
     constructor(options?: MiniCssExtractPlugin.PluginOptions);
+
+    apply(compiler: Compiler): void;
 }
 
 declare namespace MiniCssExtractPlugin {
@@ -45,7 +47,7 @@
          * This is particularly useful when dealing with multiple entry points
          * and wanting to get more control out of the filename for a given entry point/chunk
          */
-        moduleFilename?: (chunk: compilation.Chunk) => string;
+        moduleFilename?: (chunk: Chunk) => string;
     }
 }
 

--- a/node_modules/@types/duplicate-package-checker-webpack-plugin/index.d.ts	2020-05-26 14:54:57.000000000 -0700
+++ b/node_modules/@types/duplicate-package-checker-webpack-plugin/index.d.ts	2020-05-26 14:51:00.000000000 -0700
@@ -4,12 +4,13 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.3
 
-import { Plugin } from 'webpack';
+import { Compiler } from 'webpack';
 
 export = DuplicatePackageCheckerWebpackPlugin;
 
 declare class DuplicatePackageCheckerWebpackPlugin extends Plugin {
     constructor(options?: DuplicatePackageCheckerWebpackPlugin.Options);
+    apply(compiler: Compiler): void;
 }
 
 declare namespace DuplicatePackageCheckerWebpackPlugin {
